src/main/kotlin/org/rust/cargo/runconfig/buildtool/RsBuildEventsConverter.kt

```diff
@@ -3,84 +3,69 @@
  * found in the LICENSE file.
  */
 
+@file:Suppress("UnstableApiUsage")
+
 package org.rust.cargo.runconfig.buildtool
 
+import com.google.common.annotations.VisibleForTesting
 import com.google.gson.JsonObject
-import com.google.gson.JsonParser
-import com.google.gson.JsonSyntaxException
 import com.intellij.build.FilePosition
 import com.intellij.build.events.BuildEvent
 import com.intellij.build.events.MessageEvent
 import com.intellij.build.events.StartEvent
 import com.intellij.build.events.impl.*
 import com.intellij.build.output.BuildOutputInstantReader
 import com.intellij.build.output.BuildOutputParser
+import com.intellij.execution.process.AnsiEscapeDecoder
 import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.webcore.util.JsonUtil.tryParseJsonObject
+import org.rust.cargo.runconfig.RsAnsiEscapeDecoder.Companion.quantizeAnsiColors
+import org.rust.cargo.runconfig.removeEscapeSequences
 import org.rust.cargo.toolchain.impl.CargoMetadata
 import org.rust.cargo.toolchain.impl.CargoTopMessage
 import org.rust.cargo.toolchain.impl.RustcMessage
-import java.nio.file.Path
 import java.nio.file.Paths
 import java.util.function.Consumer
 
-@Suppress("UnstableApiUsage")
-class CargoBuildEventsConverter(private val context: CargoBuildContext) : BuildOutputParser {
+class RsBuildEventsConverter(private val context: CargoBuildContext) : BuildOutputParser {
+    private val decoder: AnsiEscapeDecoder = AnsiEscapeDecoder()
     private val startEvents: MutableList<StartEvent> = mutableListOf()
     private val messageEvents: MutableSet<MessageEvent> = hashSetOf()
 
-    private var jsonBuffer: String = ""
-
-    private val rawBinaries: MutableSet<String> = hashSetOf()
-    val binaries: List<Path> get() = rawBinaries.map { Paths.get(it) }
+    private val jsonBuffer: StringBuilder = StringBuilder()
 
     override fun parse(
         line: String,
         reader: BuildOutputInstantReader,
         messageConsumer: Consumer<in BuildEvent>
     ): Boolean {
-        val text = jsonBuffer + line
-        if (text.startsWith("{")) {
-            if (text.endsWith("}")) {
-                jsonBuffer = ""
-            } else {
-                jsonBuffer = text + "\n"
-                return true
-            }
+        if (jsonBuffer.isEmpty() && "{\"reason\"" !in line) {
+            return tryHandleCargoMessage(line, messageConsumer)
         }
 
-        return try {
-            val json = parseJsonObject(text)
-            tryHandleRustcMessage(json, messageConsumer) || tryHandleRustcArtifact(json)
-        } catch (e: JsonSyntaxException) {
-            tryHandleCargoMessage(text, messageConsumer)
+        jsonBuffer.append(line.withNewLine())
+        val message = jsonBuffer.dropWhile { it != '{' }.toString()
+        val jsonObject = tryParseJsonObject(message)
+        if (jsonObject == null) {
+            return false
+        } else {
+            jsonBuffer.clear()
```

Comment:
This code (and its previous version) looks tricky. Could you add a comment with a short explanation, please?